pyccel/codegen/printing/cwrappercode.py

```diff
@@ -273,88 +258,146 @@ def get_cast_function_call(self, cast_type, arg):
+
+        error = PyErr_SetString('PyExc_TypeError', self._generate_TypeError_message(variable))
+
+        return numpy_type_check, python_type_check, error
+
     # -------------------------------------------------------------------
     # Functions managing  the creation of wrapper body
     # -------------------------------------------------------------------
-    def _body_optional_variable(self, tmp_variable, variable, collect_var, check_type = False):
+
+    def _valued_variable_management(self, variable, collect_var, tmp_variable):
         """
-        Responsible for collecting value and managing error and create the body
-        of optional arguments in format
-                if (pyobject == Py_None){
-                    collect Null
-                }else if(Type Check == False){
-                    Print TypeError Wrong Type
-                    return Null
-                }else{
-                    assign pyobject value to tmp variable
-                    collect the adress of the tmp variable
-                }
+        Responsible for creating the body collecting the default value of an valuedVariable
+        and the check needed.
+        if the valuedVariable is optional create body to collect the new value
+
         Parameters:
         ----------
         tmp_variable : Variable
             The temporary variable  to hold result
-        Variable : Variable
+        variable     : Variable
             The optional variable
-        collect_var : variable
+        collect_var  : Variable
             the pyobject type variable  holder of value
-        check_type : Boolean
-            True if the type is needed
 
         Returns
         -------
-        body : list
-            A list of statements
+        section      :
+            IfSection
+        collect_body :
+            body that collect new optional variable value
         """
-        body = [IfSection(PyccelEq(VariableAddress(collect_var), VariableAddress(Py_None)),
-                [Assign(VariableAddress(variable), Nil())])]
-        if check_type : # Type check
-            check = PyccelNot(PyccelOr(NumpyType_Check(variable, collect_var)
-                    , PythonType_Check(variable, collect_var)))
-            error = PyErr_SetString('PyExc_TypeError', '"{} must be {}"'.format(variable, variable.dtype))
-            body += [IfSection(check, [error, Return([Nil()])])]
-        body += [IfSection(LiteralTrue(), [self._create_collecting_value_body(variable, collect_var, tmp_variable),
-                    Assign(VariableAddress(variable), VariableAddress(tmp_variable))])]
-        body = [If(*body)]
 
-        return body
+        valued_var_check  = PyccelEq(VariableAddress(collect_var), VariableAddress(Py_None))
+        collect_body      = []
+
+        if variable.is_optional:
+            collect_body  = [Assign(VariableAddress(variable), VariableAddress(tmp_variable))]
```

Comment:
Should this be an `AliasAssign`?

Reply:
variable is not a pointer

Reply:
Then how are you assigning to its address?

Is it not a pointer as it is an optional function argument in the c function? But here it is a local variable so I think it probably should be. This probably means that we did this poorly last time ^^'. Maybe we should clone the arguments at the start and change optional variables to pointers? Optional variables which are not arguments in their context don't make sense

Reply:
hmm will it be ok to put it here and update the comment ?
https://github.com/pyccel/pyccel/blob/c36ca7071eaed1c6775d87308acf618092fd3197/pyccel/codegen/printing/cwrappercode.py#L796

Reply:
Yes, that is exactly the right place for it :+1:

Reply:
I don't know if you can do it in that list without making it unreadable though, you might need to just add a second list

Reply:
yes i did it with a second list

Reply:
done in
3665e15   333a1a6

Reply:
@rakati  is there any reason why did you use  is_pointer and not rank > 0 ?
```PYTHON
wrapper_body += [Deallocate(i) for i in local_arg_vars if i.is_pointer]
```

Reply:
:+1: that looks clearer. It just looks like you need to handle `Deallocate` for fortran (do nothing)